pkg/component/gardenerapiserver/gardener_apiserver.go

```diff
@@ -0,0 +1,205 @@
+	values         Values
+}
+
+func (g *gardenerAPIServer) Deploy(ctx context.Context) error {
+	var (
+		runtimeRegistry = managedresources.NewRegistry(operatorclient.RuntimeScheme, operatorclient.RuntimeCodec, operatorclient.RuntimeSerializer)
+	)
+
+	runtimeResources, err := runtimeRegistry.AddAllAndSerialize()
+	if err != nil {
+		return err
+	}
+
+	timeoutCtx, cancel := context.WithTimeout(ctx, TimeoutWaitForManagedResource)
+	defer cancel()
+
+	if err := managedresources.CreateForSeed(ctx, g.client, g.namespace, managedResourceNameRuntime, false, runtimeResources); err != nil {
+		return err
+	}
+	if err := managedresources.WaitUntilHealthy(timeoutCtx, g.client, g.namespace, managedResourceNameRuntime); err != nil {
+		return err
+	}
+
+	var (
+		virtualRegistry = managedresources.NewRegistry(operatorclient.VirtualScheme, operatorclient.VirtualCodec, operatorclient.VirtualSerializer)
+	)
+
+	virtualResources, err := virtualRegistry.AddAllAndSerialize()
+	if err != nil {
+		return err
+	}
+
+	timeoutCtx, cancel = context.WithTimeout(ctx, TimeoutWaitForManagedResource)
+	defer cancel()
+
+	if err := managedresources.CreateForShoot(ctx, g.client, g.namespace, managedResourceNameVirtual, managedresources.LabelValueGardener, false, virtualResources); err != nil {
+		return err
+	}
+	return managedresources.WaitUntilHealthy(timeoutCtx, g.client, g.namespace, managedResourceNameVirtual)
+}
+
+func (g *gardenerAPIServer) Destroy(ctx context.Context) error {
+	timeoutCtx, cancel := context.WithTimeout(ctx, TimeoutWaitForManagedResource)
+	defer cancel()
+
+	if err := managedresources.DeleteForShoot(ctx, g.client, g.namespace, managedResourceNameVirtual); err != nil {
+		return err
+	}
+	if err := managedresources.WaitUntilDeleted(timeoutCtx, g.client, g.namespace, managedResourceNameVirtual); err != nil {
+		return err
+	}
+
+	timeoutCtx, cancel = context.WithTimeout(ctx, TimeoutWaitForManagedResource)
+	defer cancel()
+
+	if err := managedresources.DeleteForSeed(ctx, g.client, g.namespace, managedResourceNameRuntime); err != nil {
+		return err
+	}
+	return managedresources.WaitUntilDeleted(timeoutCtx, g.client, g.namespace, managedResourceNameRuntime)
+}
+
+func (g *gardenerAPIServer) Wait(ctx context.Context) error {
+	timeoutCtx, cancel := context.WithTimeout(ctx, TimeoutWaitForManagedResource)
+	defer cancel()
+
+	if err := g.waitUntilRuntimeManagedResourceHealthyAndNotProgressing(ctx); err != nil {
+		return err
+	}
+	return managedresources.WaitUntilHealthy(timeoutCtx, g.client, g.namespace, managedResourceNameVirtual)
+}
+
+func (g *gardenerAPIServer) WaitCleanup(ctx context.Context) error {
+	timeoutCtx, cancel := context.WithTimeout(ctx, TimeoutWaitForManagedResource)
+	defer cancel()
+
+	if err := managedresources.WaitUntilDeleted(timeoutCtx, g.client, g.namespace, managedResourceNameVirtual); err != nil {
+		return err
+	}
+	return managedresources.WaitUntilDeleted(timeoutCtx, g.client, g.namespace, managedResourceNameRuntime)
```

Comment:
I don't think we necessarily need to wait in `Destroy`, I just implemented it this way for similarity reasons (because we wait in `Deploy`). `Wait` and `WaitCleanup` are to be considered as "independent", i.e., those functions should be implemented without making assumptions of how `Deploy`/`Destroy` works, wouldn't you agree?

Reply:
> hose functions should be implemented without making assumptions of how Deploy/Destroy works, wouldn't you agree?
Yes, I agree 😄 But if we don't need to wait in `Destroy`, then we probably shouldn't do it because we won't recall why it was implemented this way.